core/src/main/java/org/infinispan/persistence/support/ComposedSegmentedLoadWriteStore.java

```diff
@@ -191,15 +193,21 @@ public void deleteBatch(Iterable<Object> keys) {
    }
 
    @Override
-   public void writeBatch(Iterable<MarshalledEntry<? extends K, ? extends V>> marshalledEntries) {
-      Flowable.fromIterable(marshalledEntries)
-            // Separate out batches by segment
+   public CompletionStage<Void> writeBatch(Publisher<MarshallableEntry<? extends K, ? extends V>> publisher) {
+      CompletableFuture<Void> future = new CompletableFuture<>();
+
+      Flowable.fromPublisher(publisher)
             .groupBy(me -> keyPartitioner.getSegment(me.getKey()))
-            .blockingForEach(groupedFlowable ->
+            .serialize()
+            .doOnNext(groupedFlowable ->
                   groupedFlowable
+                        .subscribeOn(scheduler)
                         .buffer(configuration.maxBatchSize())
-                        .blockingForEach(batch -> stores.get(groupedFlowable.getKey()).writeBatch(batch))
-            );
+                        .doOnNext(batch -> stores.get(groupedFlowable.getKey()).writeBatch(Flowable.fromIterable(batch)))
+                        .subscribe(Void -> {}, future::completeExceptionally, () -> future.complete(null))
+                  )
+            .subscribe(Void -> {}, future::completeExceptionally, () -> future.complete(null));
```

Comment:
This is more like what I wanted to see :). You can use the static method for the empty consumer instead though.